File: V1/c_v1/se2024-a2.c/production/SwarmBots_Project/src/starter.c
/**
 * \file  starter.c
 * \version  1.1
 * \author Joshua MONTREUIL
 * \date Apr 30, 2023
 * \brief entry of the program.
 *
 * \section License
 *
 * The MIT License
 *
 * Copyright (c) 2023, Prose A2 2023
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * \copyright Prose A2 2023
 * 
 */

/* ----------------------  INCLUDES  ---------------------------------------- */
#include <stdio.h>

#include <unistd.h>
#include <stdint.h>
#include <termios.h>
#include <stdlib.h>

#include "controller/state_indicator.h"
#include "controller/controller_core.h"
#include "controller/pilot.h"
#include "controller/controller_ringer.h"
#include "com/postman.h"
#include "com/dispatcher.h"
#include "logs/controller_logger.h"
#include "lib/defs.h"
/* ----------------------  PRIVATE CONFIGURATIONS  -------------------------- */
/* ----------------------  PRIVATE TYPE DEFINITIONS  ------------------------ */
/* ----------------------  PRIVATE STRUCTURES  ------------------------------ */
/* ----------------------  PRIVATE ENUMERATIONS  ---------------------------- */
/**
 * \enum log_key_e
 * \brief Defines the keys that can be used.
 */
typedef enum {
    LOG_STOP = 'q',
} log_key_e;
/* ----------------------  PRIVATE FUNCTIONS PROTOTYPES  -------------------- */
/**
 * \fn static void STARTER_capture_choice(void)
 * \brief Capture the user key entry.
 * \author Joshua MONTREUIL.
 */
static void STARTER_capture_choice(void);
/**
 * \fn static void STARTER_display(void)
 * \brief Starts the display of ihm.
 * \author Joshua MONTREUIL.
 */
static void STARTER_display(void);
/**
 * \fn static void STARTER_stop_all(void)
 * \brief Stops the app.
 * \author Joshua MONTREUIL.
 */
static void STARTER_stop_all(void);
/* ----------------------  PRIVATE VARIABLES  ------------------------------- */
/**
 * \var static bool_e quit_case
 * Used to quit the app.
 */
static bool_e quit_case = FALSE;
/* ----------------------  PUBLIC FUNCTIONS  -------------------------------- */
/* ----------------------  PRIVATE FUNCTIONS  ------------------------------- */
int main (int argc, char * argv[])
{
	printf("Hello swarmbots\n\n");
    /* MODULE CREATION */
    if(CONTROLLER_LOGGER_create() == -1) {
        printf("ERROR on controller_logger creation.\n");
        return -1;
    }
    if(STATE_INDICATOR_create() == -1) {
        printf("ERROR on state_indicator creation.\n");
        goto error_state_indicator_create;
    }
    if(CONTROLLER_CORE_create() == -1) {
        printf("ERROR on controller_core creation.\n");
        goto error_controller_core_create;
    }
    if(CONTROLLER_RINGER_create() == -1) {
        printf("ERROR on controller_ringer creation.\n");
        goto error_controller_ringer_create;
    }
    if(PILOT_create() == -1) {
        printf("ERROR on pilot creation.\n");
        goto error_pilot_create;
    }
    if(POSTMAN_create() == -1) {
        printf("ERROR on postman creation.\n");
        goto error_postman_create;
    }
    if(DISPATCHER_create() == -1) {
        printf("ERROR on postman creation.\n");
        goto error_dispatcher_create;
    }
    /* MODULE START */
    if(CONTROLLER_LOGGER_start() == -1) {
        printf("ERROR on controller logger start.\n");
        goto error_controller_logger_start;
    }
    if(STATE_INDICATOR_start() == -1) {
        printf("ERROR on state indicator start.\n");
        goto error_state_indicator_start;
    }
    if(CONTROLLER_CORE_start() == -1) {
        printf("ERROR on controller core start. \n");
        goto error_controller_core_start;
    }
    if(CONTROLLER_RINGER_start() == -1) {
        printf("ERROR on controller ringer start. \n");
        goto error_controller_ringer_start;
    }
    if(PILOT_start() == -1) {
        printf("ERROR on pilot start.\n");
        goto error_pilot_start;
    }
    if(POSTMAN_start() == -1) {
        printf("ERROR on postman start.\n");
        goto error_postman_start;
    }
    if(DISPATCHER_start() == -1) {
        printf("ERROR on postman start.\n");
        goto error_dispatcher_start;
    }
    while(!quit_case) {
        STARTER_display();
    }
    return 0;

error_dispatcher_start :
    POSTMAN_stop();
error_postman_start :
    PILOT_stop();
error_pilot_start :
    CONTROLLER_RINGER_stop();
error_controller_ringer_start :
    CONTROLLER_CORE_stop();
error_controller_core_start :
    STATE_INDICATOR_stop();
error_state_indicator_start :
    CONTROLLER_LOGGER_stop();
error_controller_logger_start :
    DISPATCHER_destroy();
error_dispatcher_create :
    POSTMAN_destroy();
error_postman_create :
    PILOT_destroy();
error_pilot_create :
    CONTROLLER_RINGER_destroy();
error_controller_ringer_create :
    CONTROLLER_CORE_destroy();
error_controller_core_create :
    STATE_INDICATOR_destroy();
error_state_indicator_create :
    CONTROLLER_LOGGER_destroy();
    return -1;
}

static void STARTER_capture_choice(void) {
    struct termios oldt, newt;
    tcgetattr(STDIN_FILENO, &oldt);
    newt = oldt;
    newt.c_lflag &= ~(ICANON | ECHO);
    tcsetattr(STDIN_FILENO, TCSANOW, &newt);
    log_key_e command = getchar();
    tcsetattr(STDIN_FILENO, TCSANOW, &oldt);

    switch(command) {
        case LOG_STOP:
        {
            STARTER_stop_all();
            break;
        }
        default:
        {
            break;
        }
    }
}

static void STARTER_display(void) {
    printf("-------------- SWARMBOTS PROG -----------------\n");
    printf("--------- Press 'q' to stop the app. --------- \n");
    STARTER_capture_choice();
}

static void STARTER_stop_all(void) {
    printf("Bye swarmbots\n\n");

    /* MODULE STOP */
    if(DISPATCHER_stop() == -1) {
        printf("ERROR on dispatcher stop.\n");
    }
    if(POSTMAN_stop() == -1) {
        printf("ERROR on postman stop.\n");
    }
    if(PILOT_stop() == -1) {
        printf("ERROR on pilot stop.\n");
    }
    if(CONTROLLER_RINGER_stop() == -1) {
        printf("ERROR on controller ringer stop. \n");
    }
    if(CONTROLLER_CORE_stop() == -1) {
        printf("ERROR on controller core stop.\n");
    }
    if(STATE_INDICATOR_stop() == -1) {
        printf("ERROR on state indicator stop.\n");
    }
    if(CONTROLLER_LOGGER_stop() == -1) {
        printf("ERROR on controller logger stop.\n");
    }
    /* MODULE DESTROY */
    if(DISPATCHER_destroy() == -1) {
        printf("ERROR on dispatcher destroy.\n");
    }
    if(POSTMAN_destroy() == -1) {
        printf("ERROR on postman destroy.\n");
    }if(PILOT_destroy() == -1) {
        printf("ERROR on pilot destroy.\n");
    }
    if(CONTROLLER_RINGER_destroy() == -1) {
        printf("ERROR on controller ringer destroy. \n");
    }
    if(CONTROLLER_CORE_destroy() == -1) {
        printf("ERROR on controller core destroy. \n");
    }
    if(STATE_INDICATOR_destroy() == -1) {
        printf("ERROR on state indicator destroy. \n");
    }
    if(CONTROLLER_LOGGER_destroy() == -1) {
        printf("ERROR on controller logger destroy. \n");
    }

    printf("END OK \n");
    quit_case = TRUE;
}

File: V1/android_v1/se2024-a2.android/production/ProjectBase/app/src/main/java/project/src/model/core/Starter.java
/**
 * \file  Starter.java
 * \version  0.1
 * \author Mahery FONG
 * \date May 18, 2023
 * \brief Starter is responsible for the start of the application and active object threads.
 * It also detects the application stop and stop all active object threads.
 *
 * \section License
 *
 * The MIT License
 *
 * Copyright (c) 2023, Prose A2 2023
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * \copyright Prose A2 2023
 *
 */

/* ----------------------  PACKAGE  ---------------------------------------- */
package project.src.model.core;

/* ----------------------  INCLUDES  ---------------------------------------- */

import java.util.concurrent.ArrayBlockingQueue;

public class Starter extends Thread {

    /* ----------------------  PRIVATE CONSTANT  ------------------------------ */

    /**
     * @brief Number of event accepted by the GUISecretary queue.
     */
    private static final int MAX_CAPACITY_QUEUE = 10;

    /* ----------------------  PRIVATE OBJECTS  ------------------------------ */

    /**
     * @brief State of the application.
     */
    public enum ApplicationState {
        OPENED,
        CLOSED
    }

    /**
     * @brief State of the screens.
     */
    public enum ScreenState {
        STARTED,
        STOPPED
    }

    /**
     * @brief Starter queue to send messages.
     */
    private final ArrayBlockingQueue<Event> starterQueue;

    /* ----------------------  PRIVATE VARIABLES  ---------------------------- */

    /**
     * @brief Current state of the application.
     */
    private ApplicationState applicationState = ApplicationState.OPENED;

    /**
     * @brief Current state of the home screen.
     */
    private ScreenState homeScreenState = ScreenState.STOPPED;

    /**
     * @brief Current state of the command screen.
     */
    private ScreenState commandScreenState = ScreenState.STOPPED;

    /**
     * @brief Current state of the logs screen.
     */
    private ScreenState logsScreenState = ScreenState.STOPPED;

    /**
     * @brief Instance of the starter basic singleton object.
     */
    public static Starter INSTANCE;

    static {
        INSTANCE = new Starter();
    }

    /**
     * @brief Private constructor for the basic singleton Starter object.
     */
    private Starter() {
        this.starterQueue = new ArrayBlockingQueue<>(MAX_CAPACITY_QUEUE);
    }

    /**
     * @brief Create and start the active objects.
     */
    public void main() {
        this.homeScreenState = ScreenState.STARTED;

        GUI.INSTANCE.create();
        GUISecretary.INSTANCE.create();

        if (!LogsManager.currentThread().isDaemon()) LogsManager.INSTANCE.start();
        if (!GUISecretary.currentThread().isDaemon()) GUISecretary.INSTANCE.start();
        if (!GUI.currentThread().isDaemon()) GUI.INSTANCE.start();
        if (!Starter.currentThread().isDaemon()) Starter.INSTANCE.start();
    }

    /**
     * @brief Main function of Starter thread. It receives the events from the other objects.
     */
    @Override
    public void run() {
        // Event store in the variable.
        Event event = null;
        while(applicationState == ApplicationState.OPENED) {

            try {
                // Catch the oldest message event in the queue.
                event = this.starterQueue.take();
            } catch (InterruptedException e) {

            }

            // Call the update function depending on the event start or stop received.
            if (event.getEventID() == Event.EventID.SCREEN_START) updateStart(event.getScreenActivity());
            else updateStopped(event.getScreenActivity());

            // Check if all screen is stopped.
            this.checkAllStopped();
        }
    }

    /**
     * @brief Notify state change of the home activity to set its new state.
     * @param screenState State of the home activity screen.
     */
    public void notifyHomeActivity(ScreenState screenState) {
        Event event = new Event( (screenState == ScreenState.STARTED) ?
                Event.EventID.SCREEN_START : Event.EventID.SCREEN_STOPPED);
        event.setScreenActivity(Event.ScreenActivity.HOME_SCREEN);
        this.starterQueue.add(event);
    }

    /**
     * @brief Notify state change of the command activity to set its new state.
     * @param screenState State of the command activity screen.
     */
    public void notifyCommandActivity(ScreenState screenState) {
        Event event = new Event( (screenState == ScreenState.STARTED) ?
                Event.EventID.SCREEN_START : Event.EventID.SCREEN_STOPPED);
        event.setScreenActivity(Event.ScreenActivity.COMMAND_SCREEN);
        this.starterQueue.add(event);
    }

    /**
     * @brief Notify state change of the logs activity to set its new state.
     * @param screenState State of the logs activity screen.
     */
    public void notifyLogsActivity(ScreenState screenState) {
        Event event = new Event( (screenState == ScreenState.STARTED) ?
                Event.EventID.SCREEN_START : Event.EventID.SCREEN_STOPPED);
        event.setScreenActivity(Event.ScreenActivity.LOGS_SCREEN);
        this.starterQueue.add(event);
    }

    /**
     * @brief Check the state of all screen if they're stopped. If yes, engages the disconnection of
     * robots, and stops active objects and application.
     */
    private void checkAllStopped() {
        if ( (this.homeScreenState == ScreenState.STOPPED)
                & (this.commandScreenState == ScreenState.STOPPED)
                & (this.logsScreenState == ScreenState.STOPPED) )
        {
            // Stop application detected.
            GUI.INSTANCE.quit();
            this.applicationState = ApplicationState.CLOSED;
        }
    }

    /**
     * @brief Set the state of the given screen to running.
     * @param screen Screen concerned by the state change.
     */
    private void updateStart(Event.ScreenActivity screen) {
        switch (screen) {
            case HOME_SCREEN :
                this.homeScreenState = ScreenState.STARTED;
                break;
            case COMMAND_SCREEN :
                this.commandScreenState = ScreenState.STARTED;
                break;
            case LOGS_SCREEN :
                this.logsScreenState = ScreenState.STARTED;
                break;
            default :
                break;
        }
    }

    /**
     * @brief Set the state of the given screen to stopped.
     * @param screen Screen concerned by the state change.
     */
    private void updateStopped(Event.ScreenActivity screen) {
        switch (screen) {
            case HOME_SCREEN :
                this.homeScreenState = ScreenState.STOPPED;
                break;
            case COMMAND_SCREEN :
                this.commandScreenState = ScreenState.STOPPED;
                break;
            case LOGS_SCREEN :
                this.logsScreenState = ScreenState.STOPPED;
                break;
            default :
                break;
        }
    }

}
